README.md
# CODEF API - Python
Python sample for CODEF API

## Documentation

본 샘플은 CODEF API의 연동을 위한 공통 코드를 포함하고 있으며, 지원하는 모든 API의 엔드포인트(은행, 카드, 보험, 증권, 공공, 기타)는
https://develpers.codef.io 를 통해 확인할 수 있습니다.

## CODEF API Env

CODEF API는 원활한 개발을 위해 샌드박스, 개발, 운영 환경을 각각 제공한다.

- 샌드박스 : https://sandbox.codef.io
- 개발 : https://development.codef.io
- 운영 : https://api.codef.io

## Getting Started


### OAuth2.0

CODEF API를 사용하기 위해서는 'access_token' 발행이 선행되어야 하며, 거래 시 Header 에 포함하여 요청합니다.
'access_token'을 발급 받기 위한 'client_id' 및 'client_secret'은 https://codef.io/#/account/keys 에서 확인할 수 있습니다.

```python
token_url = 'https://api.codef.io/oauth/token'
response_oauth = request_token(token_url, "codef_master", "codef_master_secret");
    if response_oauth.status_code == 200:
        dict = json.loads(response_oauth.text)
        # reissue_token
        token = dict['access_token']

        print('access_token = ' + token)
    else:
        print('토큰발급 오류')
```
```json
{"access_token":"eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJzZXJ2aWNlX3R5cGUiOiIwIiwic2NvcGUiOlsicmVhZCJdLCJzZXJ2aWNlX25vIjoiMDAwMDAwMDQyMDAxIiwiZXhwIjoxNTYyNjc0NTczLCJhdXRob3JpdGllcyI6WyJJTlNVUkFOQ0UiLCJQVUJMSUMiLCJCQU5LIiwiRVRDIiwiU1RPQ0siLCJDQVJEIl0sImp0aSI6ImFiNTBjM2RmLWQ3MzctNGE2Ny04Zjg4LWQzOTE2YTNiYmNiMSIsImNsaWVudF9pZCI6ImNvZGVmX21hc3RlciJ9.EXBV-D89_zoYmFdiULahGqcp1T2Du8DM51Trf1fD4MxsKYsA1t37ovffIKIQvqLHwQz4W8EqC6s8lM1V_IqFG5D5yafmyvprVi7ciqRMBBIsnEZN8xk1gBqLydtwkG0jKTrCLTBls8zATHbWV8BO6oUw8fwQId4ExeewbqeflSBCLOztb4c8UkR1WFDqQs63Ezry8k79VN5HPSktChJGnGq0xWmtbMlwv8IubvveJkMLz-6Iw6hlSMjeat_fv-gZCPTPdoaMa-BPxcAhI772cSCrfJNzori0uVFIeBEInabDzAKpXjvbsZEz_q70QGGSPkoslxFb_N-MYSNPgCWEvw","token_type":"bearer","expires_in":9,"scope":"read"}
```


### 계정 생성

CODEF API를 사용하기 위해서는 엔드유저가 사용하는 대상기관의 인증수단 등록이 필요하며, 이를 통해 사용자마다 유니크한 'connected_id'를 발급받을 수 있습니다.
이후에는 별도의 인증수단 전송 없이 'connected_id'를 통해서 대상기관의 데이터를 연동할 수 있습니다.

```python
codef_account_create_url = 'https://api.codef.io/account/create'
codef_account_create_body = {
            'accountList':[                    # 계정목록
                {
                    'countryCode':'KR',        # 국가코드
                    'businessType':'BK',       # 비즈니스 구분
                    'clientType':'P',          # 고객구분(P: 개인, B: 기업)
                    'organization':'0004',     # 기관코드
                    'loginType':'0',           # 로그인타입 (0: 인증서, 1: ID/PW)
                    'password':'1234',         # 인증서 비밀번호             
                    'derFile':'인증서 DerFile',  # Base64String
                    'keyFile':'인증서 KeyFile'   # Base64String
                }
            ]
}

# CODEF API 호출
response_account_create = http_sender(codef_account_create_url, token, codef_account_create_body)
dict = json.loads(urllib.unquote_plus(response.text.encode('utf8')))
connected_id = dict['data']['connectedId']
```
```json
{"result":{"code":"CF-00000","extraMessage":"","message":"정상"},"data":{"organizationList":[{"clientType":"P","loginType":"0","organization":"0004","businessType":"BK"}],"connectedId":"dE-HSO244D-8Yf2nsMLunI"}}
```


### 계정 추가

계정 생성을 통해 발급받은 'connected_id'에 추가 기관의 인증수단을 등록할 수 있습니다. 추가 등록한 기관을 포함하여 이후에는 별도의 인증수단 전송없이
'connected_id'를 통해서 대상기관의 데이터를 연동할 수 있습니다.

```python
codef_account_add_url = 'https://api.codef.io/account/add'
codef_account_add_body = {
            'connectedId': '계정생성 시 발급받은 아이디',    # connected_id
            'accountList':[                    # 계정목록
                {
                    'countryCode':'KR',        # 국가코드
                    'businessType':'BK',       # 비즈니스 구분
                    'clientType':'P',          # 고객구분(P: 개인, B: 기업)
                    'organization':'0020',     # 기관코드
                    'loginType':'0',           # 로그인타입 (0: 인증서, 1: ID/PW)
                    'derFile':'인증서 DerFile',  # Base64String
                    'keyFile':'인증서 KeyFile'   # Base64String
                }
            ]
}

# CODEF API 호출
response_account_add = http_sender(codef_account_add_url, token, codef_account_add_body)
```
```json
{"result":{"code":"CF-00000","extraMessage":"","message":"정상"},"data":{"successList":[{"clientType":"P","code":"CF-00000","countryCode":"KR","organization":"0020","businessType":"BK","message":"성공"}],"errorList":[],"connectedId":"9cbO3CatAhzbqR5JNOX5IZ"}}
```


### 계정 수정

계정 생성을 통해 발급받은 'connected_id'에 등록된 기관의 인증수단을 변경할 수 있습니다. 변경 요청한 기관의 인증 수단은 호출 즉시 변경되며, 이 후
'connected_id'를 통해서 대상기관의 데이터를 연동할 수 있습니다.

```python
codef_account_update_url = 'https://api.codef.io/account/update'
codef_account_update_body = {
            'connectedId': '계정생성 시 발급받은 아이디',    # connected_id
            'accountList':[                    # 계정목록
                {
                    'countryCode':'KR',        # 국가코드
                    'businessType':'BK',       # 비즈니스 구분
                    'clientType':'P',          # 고객구분(P: 개인, B: 기업)
                    'organization':'0020',     # 기관코드
                    'loginType':'0',           # 로그인타입 (0: 인증서, 1: ID/PW)
                    'password':'1234',         # 인증서 비밀번호             
                    'derFile':'인증서 DerFile',  # Base64String
                    'keyFile':'인증서 KeyFile'   # Base64String
                }
            ]
}

# CODEF API 호출
response_account_update = http_sender(codef_account_update_url, token, codef_account_update_body)
```
```json
{"result":{"code":"CF-00000","extraMessage":"","message":"정상"},"data":{"successList":[{"clientType":"P","code":"CF-00000","countryCode":"KR","organization":"0020","businessType":"BK","message":"성공"}],"errorList":[],"connectedId":"9cbO3CatAhzbqR5JNOX5IZ"}}
```


### 계정 삭제

엔드유저가 등록된 계정의 삭제를 요청 시 'connected_id'에 등록된 기관의 인증수단을 즉시 삭제할 수 있습니다. 요청한 기관의 인증 수단은 호출 즉시 삭제되며,
해당 데이터는 복구할 수 없습니다.

```python
codef_account_delete_url = 'https://api.codef.io/account/delete'
codef_account_delete_body = {
            'connectedId': '계정생성 시 발급받은 아이디',    # connected_id
            'accountList':[                    # 계정목록
                {
                    'countryCode':'KR',        # 국가코드
                    'businessType':'BK',       # 비즈니스 구분
                    'clientType':'P',          # 고객구분(P: 개인, B: 기업)
                    'organization':'0020',     # 기관코드
                    'loginType':'0',           # 로그인타입 (0: 인증서, 1: ID/PW)
                    'password':'1234',         # 인증서 비밀번호             
                    'derFile':'인증서 DerFile',  # Base64String
                    'keyFile':'인증서 KeyFile'   # Base64String
                }
            ]
}

# CODEF API 호출
response_account_delete = http_sender(codef_account_delete_url, token, codef_account_delete_body)
```
```json
{"result":{"code":"CF-00000","extraMessage":"","message":"정상"},"data":{"organizationList":[{"clientType":"P","loginType":"0","countryCode":"KR","organization":"0020","businessType":"BK"}],"connectedId":"9cbO3CatAhzbqR5JNOX5IZ"}}
```


### CODEF API(법인 보유계좌조회)

발급받은 'connected_id' 를 통해 등록된 기관의 보유계좌를 조회할 수 있습니다.

TestKR_BK_1_B_001.py
```python
# CodefURL
codef_url = 'https://api.codef.io'
token_url = 'https://api.codef.io/oauth/token'

# 은행 법인 보유계좌
account_list_path = '/v1/kr/bank/b/account/list'

# 기 발급된 토큰
token =''     #access_token

# BodyData
body = {
    'connected_id':'9LUm.uhVQbzaangazwI0tr',
    'organization':'0011'
}

# CODEF API 요청
response_codef_api = http_sender(codef_url + account_list_path, token, body)

# token error
if response_codef_api.status_code == 401:
    dict = json.loads(response_codef_api.text)
    # invalid_token
    print('error = ' + dict['error'])
    # Cannot convert access token to JSON
    print('error_description = ' + dict['error_description'])

    # reissue token
    response_oauth = request_token(token_url, "codef_master", "codef_master_secret");
    if response_oauth.status_code == 200:
        dict = json.loads(response_oauth.text)
        # reissue_token
        token = dict['access_token']
        print('access_token = ' + token)

        # request codef_api
        response = http_sender(codef_url + account_list_path, token, body)

        # codef_api 응답 결과
        print(response.status_code)
        print(response.text)
    else:
        print('토큰발급 오류')
else:
    print('정상처리')
```
```json
{"result":{"code":"CF-94002","extraMessage":"","message":"사용자+계정정보+설정에+실패했습니다."},"data":{}}
```


### 오류

CODEF API 오류는 HTTP status code 와 CODEF API ErrorCode로 분류합니다.

HTTP 401 - OAuth2.0 토큰 만료
```json
{"error":"invalid_token","error_description":"Cannot convert access token to JSON","code":"CF-99997","message":"OAUTH2.0 토큰 에러입니다. 메시지를 확인하세요."}
```

그 외 오류 HTTP 200 - CODEF 오류 변환(CF-XXXXX)
```json
{"result":{"code":"CF-94002","extraMessage":"","message":"사용자 계정정보 설정에 실패했습니다."},"data":{}}
```


## Change Log

CODEF API의 변경내을 [CHANGELOG.md](CHANGELOG.md) 을 통해 확인할 수 있습니다.

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details


CHANGELOG.md
## 1.0.0(2019-07-10)
* CODEF API 등록
  * 계정등록, 은행, 카드 API 등록


RegisterAccount.py
# -*- coding: utf-8 -*-
# UTF-8 encoding when using korean
#
import requests, json, base64
import urllib

# ========== HTTP 기본 함수 ==========

def http_sender(url, token, body):
    headers = {'Content-Type': 'application/json',
        'Authorization': 'Bearer ' + token
        }

    response = requests.post(url, headers = headers, data = json.dumps(body))

    print('response.status_code = ' + str(response.status_code))
    print('response.text = ' + urllib.unquote_plus(response.text.encode('utf8')))

    return response
# ========== HTTP 함수  ==========

# ========== Toekn 재발급  ==========
def request_token(url, client_id, client_secret):
    authHeader = stringToBase64(client_id + ':' + client_secret)

    headers = {
        'Acceppt': 'application/json',
        'Content-Type': 'application/x-www-form-urlencoded',
        'Authorization': 'Basic ' + authHeader
        }

    response = requests.post(url, headers = headers, data = 'grant_type=client_credentials&scope=read')

    print(response.status_code)
    print(response.text)

    return response
# ========== Toekn 재발급  ==========

# ========== Encode string data  ==========
def stringToBase64(s):
    return base64.b64encode(s.encode('utf-8'))

def base64ToString(b):
    return base64.b64decode(b).decode('utf-8')
# ========== Encode string data  ==========

# token URL
token_url = 'https://toauth.codef.io/oauth/token'

# CODEF 연결 아이디
connected_id = ''

# 기 발급된 토큰
token ='eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJzZXJ2aWNlX3R5cGUiOiIwIiwic2NvcGUiOlsicmVhZCJdLCJzZXJ2aWNlX25vIjoiOTk5OTk5OTk5OSIsImV4cCI6MTU2MjczNjUyNywiYXV0aG9yaXRpZXMiOlsiSU5TVVJBTkNFIiwiUFVCTElDIiwiQkFOSyIsIkVUQyIsIlNUT0NLIiwiQ0FSRCJdLCJqdGkiOiIyODBhNjVkOS02NjU1LTQ5MzYtODEwNS05MjUyYTk4MGRjMDgiLCJjbGllbnRfaWQiOiJjb2RlZl9tYXN0ZXIifQ.eFCEgxcntsEkjFORAWGSi6949UMOuCxVsm2wnYlDXqrHWXXwG7-XfKugsBNone_qRRGeKD3iv6f_TEcVMWyTz8aS0fRbE514LVz6PnzKbruyPNDA5Pk3ym8up9h4Ba1ix__Bvpu_TB0Y7Fikk9YHWHacJy4F_WOjr8xFP-q2egh763_LqVUzRakGQoLOTukduZ5zH5lfSO1Z9yx2cnDkY4VSM9DTbycSZuA2oQkMVpXJc0slEyWLw7WNX5E-ff3fL6ePfJvu7by_4KmgmmJkOoKBWvJ00DwrwhAa1EZmjqGPYG6RE6wxSwsu3lYeiCX-jSGm_cbKdk7YDnYxm8FKzg'

##############################################################################
##                               계정 생성 Sample                             ##
##############################################################################
# Input Param
#
# accountList : 계정목록
#   countryCode : 국가코드
#   businessType : 비즈니스 구분
#   clientType : 고객구분(P: 개인, B: 기업)
#   organization : 기관코드
#   loginType : 로그인타입 (0: 인증서, 1: ID/PW)
#   password : 인증서 비밀번호
#   derFile : 인증서 derFile
#   keyFile : 인증서 keyFile
#
##############################################################################
print('=============================== 계정생성 ===============================')
codef_account_create_url = 'https://tapi.codef.io/v1/account/create'
codef_account_create_body = {
            'accountList':[
                {
                    'countryCode':'KR',
                    'businessType':'BK',
                    'clientType':'P',
                    'organization':'0004',
                    'loginType':'0',
                    'password':'1234',      # 인증서 비밀번호 입력
                    'derFile':'MIIF...',    # 인증서 인증서 DerFile
                    'keyFile':'MIIF...'     # 인증서 인증서 KeyFile
                }
            ]
}

response_account_create = http_sender(codef_account_create_url, token, codef_account_create_body)
if response_account_create.status_code == 200:      # success
    dict = json.loads(urllib.unquote_plus(response_account_create.text.encode('utf8')))
    if 'data' in dict and str(dict['data']) != '{}':
        data = dict['data']
        if 'connectedId' in data:
            connected_id = data['connectedId']
            print('connected_id = ' + connected_id)
            print('계정생성 정상처리')
    else:
        print(urllib.unquote_plus(response_account_create.text.encode('utf8')))
        print('계정생성 오류')
elif response_account_create.status_code == 401:      # token error
    dict = json.loads(response_account_create.text)
    # invalid_token
    print('error = ' + dict['error'])
    # Cannot convert access token to JSON
    print('error_description = ' + dict['error_description'])

    # reissue token
    response_oauth = request_token(token_url, "codef_master", "codef_master_secret");
    if response_oauth.status_code == 200:
        dict = json.loads(response_oauth.text)
        # reissue_token
        token = dict['access_token']
        print('access_token = ' + token)

        # request codef_api
        response = http_sender(codef_account_create_url, token, codef_account_create_body)
        if response.status_code == 200:      # success
            dict = json.loads(urllib.unquote_plus(response.text.encode('utf8')))
            if 'data' in dict and str(dict['data']) != '{}':
                data = dict['data']
                if 'connectedId' in data:
                    connected_id = data['connectedId']
                    print('connected_id = ' + connected_id)
                    print('계정생성 정상처리')
            else:
                print(response.text)
    else:
        print('토큰발급 오류')
else:
    print('계정생성 오류')



##############################################################################
##                               계정 추가 Sample                             ##
##############################################################################
# Input Param
#
# connectedId : CODEF 연결아이디
# accountList : 계정목록
#   countryCode : 국가코드
#   businessType : 비즈니스 구분
#   clientType : 고객구분(P: 개인, B: 기업)
#   organization : 기관코드
#   loginType : 로그인타입 (0: 인증서, 1: ID/PW)
#   password : 인증서 비밀번호
#   derFile : 인증서 derFile
#   keyFile : 인증서 keyFile
#
##############################################################################
print('=============================== 계정추가 ===============================')
codef_account_add_url = 'https://tapi.codef.io/v1/account/add'
codef_account_add_body = {
            'connectedId': '8-cXc.6lk-ib4Whi5zClVt',    # connected_id
            'accountList':[
                {
                    'countryCode':'KR',
                    'businessType':'BK',
                    'clientType':'P',
                    'organization':'0020',
                    'loginType':'0',
                    'password':'1234',      # 인증서 비밀번호 입력
                    'derFile':'MIIF...',    # 인증서 인증서 DerFile
                    'keyFile':'MIIF...'     # 인증서 인증서 KeyFile
                }
            ]
}

response_account_add = http_sender(codef_account_add_url, token, codef_account_add_body)
if response_account_add.status_code == 200:        # token error
    print('계정추가 정상처리')
elif response_account_add.status_code == 401:      # token error
    dict = json.loads(response_account_add.text)
    # invalid_token
    print('error = ' + dict['error'])
    # Cannot convert access token to JSON
    print('error_description = ' + dict['error_description'])

    # reissue token
    response_oauth = request_token(token_url, "codef_master", "codef_master_secret");
    if response_oauth.status_code == 200:
        dict = json.loads(response_oauth.text)
        # reissue_token
        token = dict['access_token']
        print('access_token = ' + token)

        # request codef_api
        response = http_sender(codef_account_add_url, token, codef_account_add_body)
        print('계정추가 정상처리')

        # codef_api 응답 결과
        print(response.status_code)
        print(response.text)
    else:
        print('토큰발급 오류')
else:
    print('계정추가 정상처리')



##############################################################################
##                               계정 수정 Sample                             ##
##############################################################################
# Input Param
#
# connectedId : CODEF 연결아이디
# accountList : 계정목록
#   countryCode : 국가코드
#   businessType : 비즈니스 구분
#   clientType : 고객구분(P: 개인, B: 기업)
#   organization : 기관코드
#   loginType : 로그인타입 (0: 인증서, 1: ID/PW)
#   password : 인증서 비밀번호
#   derFile : 인증서 derFile
#   keyFile : 인증서 keyFile
#
##############################################################################
print('=============================== 계정수정 ===============================')
codef_account_update_url = 'https://tapi.codef.io/v1/account/update'
codef_account_update_body = {
            'connectedId': '8-cXc.6lk-ib4Whi5zClVt',        # connected_id
            'accountList':[
                {
                    'countryCode':'KR',
                    'businessType':'BK',
                    'clientType':'P',
                    'organization':'0020',
                    'loginType':'0',
                    'password':'1234',      # 인증서 비밀번호 입력
                    'derFile':'MIIF...',    # 인증서 인증서 DerFile
                    'keyFile':'MIIF...'     # 인증서 인증서 KeyFile
                }
            ]
}

response_account_update = http_sender(codef_account_update_url, token, codef_account_update_body)
if response_account_update.status_code == 200:      # success
    dict = json.loads(urllib.unquote_plus(response_account_update.text.encode('utf8')))
    if 'data' in dict and str(dict['data']) != '{}':
        data = dict['data']
        if 'connectedId' in data:
            connected_id = data['connectedId']
            print('connected_id = ' + connected_id)
            print('계정수정 정상처리')
    else:
        print(response_account_update.text)
elif response_account_update.status_code == 401:      # token error
    dict = json.loads(response_account_update.text)
    # invalid_token
    print('error = ' + dict['error'])
    # Cannot convert access token to JSON
    print('error_description = ' + dict['error_description'])

    # reissue token
    response_oauth = request_token(token_url, "codef_master", "codef_master_secret");
    if response_oauth.status_code == 200:
        dict = json.loads(response_oauth.text)
        # reissue_token
        token = dict['access_token']
        print('access_token = ' + token)

        # request codef_api
        response = http_sender(codef_account_update_url, token, codef_account_update_body)
        if response.status_code == 200:      # success
            dict = json.loads(urllib.unquote_plus(response.text.encode('utf8')))
            if 'data' in dict and str(dict['data']) != '{}':
                data = dict['data']
                if 'connectedId' in data:
                    connected_id = data['connectedId']
                    print('connected_id = ' + connected_id)
                    print('계정수정 정상처리')
            else:
                print(response.text)
    else:
        print('토큰발급 오류')
else:
    print('계정수정 오류')



##############################################################################
##                               계정 삭제 Sample                             ##
##############################################################################
# Input Param
#
# connectedId : CODEF 연결아이디
# accountList : 계정목록
#   countryCode : 국가코드
#   businessType : 비즈니스 구분
#   clientType : 고객구분(P: 개인, B: 기업)
#   organization : 기관코드
#   loginType : 로그인타입 (0: 인증서, 1: ID/PW)
#   password : 인증서 비밀번호
#   derFile : 인증서 derFile
#   keyFile : 인증서 keyFile
#
##############################################################################
print('=============================== 계정삭제 ===============================')
codef_account_delete_url = 'https://tapi.codef.io/v1/account/delete'
codef_account_delete_body = {
            'connectedId': '8-cXc.6lk-ib4Whi5zClVt',        # connected_id
            'accountList':[
                {
                    'countryCode':'KR',
                    'businessType':'BK',
                    'clientType':'P',
                    'organization':'0020',
                    'loginType':'0',
                    'password':'1234',      # 인증서 비밀번호 입력
                    'derFile':'MIIF...',    # 인증서 인증서 DerFile
                    'keyFile':'MIIF...'     # 인증서 인증서 KeyFile
                }
            ]
}

response_account_delete = http_sender(codef_account_delete_url, token, codef_account_delete_body)
if response_account_delete.status_code == 200:      # success
    dict = json.loads(urllib.unquote_plus(response_account_delete.text.encode('utf8')))
    if 'data' in dict and str(dict['data']) != '{}':
        data = dict['data']
        if 'connectedId' in data:
            connected_id = data['connectedId']
            print('connected_id = ' + connected_id)
            print('계정삭제 정상처리')
    else:
        print(response_account_delete.text)
elif response_account_delete.status_code == 401:      # token error
    dict = json.loads(response_account_delete.text)
    # invalid_token
    print('error = ' + dict['error'])
    # Cannot convert access token to JSON
    print('error_description = ' + dict['error_description'])

    # reissue token
    response_oauth = request_token(token_url, "codef_master", "codef_master_secret");
    if response_oauth.status_code == 200:
        dict = json.loads(response_oauth.text)
        # reissue_token
        token = dict['access_token']
        print('access_token = ' + token)

        # request codef_api
        response = http_sender(codef_account_delete_url, token, codef_account_delete_body)
        if response.status_code == 200:      # success
            dict = json.loads(urllib.unquote_plus(response.text.encode('utf8')))
            if 'data' in dict and str(dict['data']) != '{}':
                data = dict['data']
                if 'connectedId' in data:
                    connected_id = data['connectedId']
                    print('connected_id = ' + connected_id)
                    print('계정삭제 정상처리')
            else:
                print(response.text)
    else:
        print('토큰발급 오류')
else:
    print('계정삭제 오류')
